File: example/modules/expo-native-request-trigger/android/src/main/java/expo/modules/nativerequesttrigger/NativeRequestTriggerModule.kt
package expo.modules.nativerequesttrigger

import android.util.Log
import expo.modules.inappdebugger.InAppDebuggerOkHttpIntegration
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.IOException
import java.util.Locale
import java.util.concurrent.TimeUnit
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response

private const val TAG = "NativeRequestTrigger"

private val nativeRequestClient: OkHttpClient by lazy {
  val builder =
    OkHttpClient.Builder()
      .connectTimeout(15, TimeUnit.SECONDS)
      .readTimeout(20, TimeUnit.SECONDS)
      .writeTimeout(20, TimeUnit.SECONDS)

  // This mirrors how a host-owned native module should wire its own OkHttp stack.
  InAppDebuggerOkHttpIntegration.instrument(builder)
  builder.build()
}

class NativeRequestTriggerModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("NativeRequestTrigger")

    AsyncFunction("sendHttpRequest") { rawOptions: Map<String, Any?> ->
      val url =
        (rawOptions["url"] as? String)
          ?.trim()
          ?.takeIf { it.isNotEmpty() }
          ?: throw IllegalArgumentException("NativeRequestTrigger.sendHttpRequest requires a non-empty url.")
      val method = ((rawOptions["method"] as? String) ?: "GET").uppercase(Locale.US)
      val body = rawOptions["body"]?.toString()
      val headers = normalizeHeaders(rawOptions["headers"] as? Map<*, *>)
      val request = buildRequest(url = url, method = method, body = body, headers = headers)

      nativeRequestClient.newCall(request).enqueue(
        object : Callback {
          override fun onFailure(call: Call, e: IOException) {
            Log.w(TAG, "Native $method failed for $url: ${e.message}", e)
          }

          override fun onResponse(call: Call, response: Response) {
            response.use {
              // Drain the body so the debugger can observe the full native request lifecycle.
              it.body?.bytes()
              Log.i(TAG, "Native $method completed for $url with status ${it.code}")
            }
          }
        }
      )
    }
  }
}

private fun normalizeHeaders(rawHeaders: Map<*, *>?): Map<String, String> {
  val headers =
    linkedMapOf(
      "Accept" to "application/json",
      "X-Debug-Native-Plugin" to "example-expo-module",
      "X-Debug-Native-Platform" to "android"
    )

  rawHeaders?.forEach { (key, value) ->
    val normalizedKey = key as? String ?: return@forEach
    headers[normalizedKey] = value?.toString() ?: ""
  }
  return headers
}

private fun buildRequest(
  url: String,
  method: String,
  body: String?,
  headers: Map<String, String>
): Request {
  val requestBuilder = Request.Builder().url(url)
  headers.forEach { (key, value) ->
    requestBuilder.header(key, value)
  }

  return when (method) {
    "GET" -> requestBuilder.get().build()
    "POST" -> {
      val contentType =
        headers.entries.firstOrNull { it.key.equals("Content-Type", ignoreCase = true) }?.value
          ?: "application/json; charset=utf-8"
      requestBuilder
        .post((body ?: "").toRequestBody(contentType.toMediaTypeOrNull()))
        .build()
    }
    else -> throw IllegalArgumentException("NativeRequestTrigger only supports GET and POST, received $method.")
  }
}

File: example/modules/expo-native-request-trigger/ios/NativeRequestTriggerModule.swift
import ExpoModulesCore
import Foundation

private enum NativeRequestTriggerMethod: String {
  case get = "GET"
  case post = "POST"
}

private enum NativeRequestTriggerClient {
  static let session: URLSession = {
    let configuration = URLSessionConfiguration.ephemeral
    configuration.timeoutIntervalForRequest = 15
    configuration.timeoutIntervalForResource = 20
    return URLSession(configuration: configuration)
  }()
}

public final class NativeRequestTriggerModule: Module {
  public func definition() -> ModuleDefinition {
    Name("NativeRequestTrigger")

    AsyncFunction("sendHttpRequest") { (rawOptions: [String: Any]) in
      let urlString =
        (rawOptions["url"] as? String)?
        .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
      guard !urlString.isEmpty, let url = URL(string: urlString) else {
        throw Exception(
          name: "ERR_NATIVE_REQUEST_TRIGGER_INVALID_URL",
          description: "NativeRequestTrigger.sendHttpRequest requires a valid url."
        )
      }

      let methodValue = ((rawOptions["method"] as? String) ?? NativeRequestTriggerMethod.get.rawValue)
        .uppercased()
      guard let method = NativeRequestTriggerMethod(rawValue: methodValue) else {
        throw Exception(
          name: "ERR_NATIVE_REQUEST_TRIGGER_UNSUPPORTED_METHOD",
          description: "NativeRequestTrigger only supports GET and POST, received \(methodValue)."
        )
      }

      let body = rawOptions["body"] as? String
      let userHeaders = rawOptions["headers"] as? [String: Any] ?? [:]
      var headers: [String: String] = [
        "Accept": "application/json",
        "X-Debug-Native-Plugin": "example-expo-module",
        "X-Debug-Native-Platform": "ios",
      ]
      userHeaders.forEach { key, value in
        headers[key] = String(describing: value)
      }

      var request = URLRequest(url: url, timeoutInterval: 20)
      request.httpMethod = method.rawValue

      if method == .post {
        request.httpBody = (body ?? "").data(using: .utf8)
        if headers.keys.first(where: { $0.caseInsensitiveCompare("Content-Type") == .orderedSame }) == nil {
          headers["Content-Type"] = "application/json; charset=utf-8"
        }
      }

      headers.forEach { key, value in
        request.setValue(value, forHTTPHeaderField: key)
      }

      NativeRequestTriggerClient.session.dataTask(with: request) { data, response, error in
        if let error {
          NSLog("[NativeRequestTrigger] Native \(method.rawValue) failed for \(urlString): \(error.localizedDescription)")
          return
        }
        let statusCode = (response as? HTTPURLResponse)?.statusCode ?? -1
        NSLog(
          "[NativeRequestTrigger] Native \(method.rawValue) completed for \(urlString) " +
            "status=\(statusCode) bytes=\(data?.count ?? 0)"
        )
      }.resume()
    }
  }
}
